ee/server/scim/users.go

```diff
@@ -436,23 +437,191 @@ func (u *UserHandler) Patch(r *http.Request, id string, operations []scim.PatchO
+					return scim.Resource{}, errors.ScimErrorBadParams([]string{fmt.Sprintf("%v", op)})
+				}
 			}
-			active, err := getRequiredResource[bool](newValues, activeAttr)
-			if err != nil {
-				level.Info(u.logger).Log("msg", "failed to get active value", "value", op.Value)
-				return scim.Resource{}, err
+		case op.Path.String() == userNameAttr:
+			userName, ok := op.Value.(string)
+			if !ok || userName == "" {
+				level.Info(u.logger).Log("msg", fmt.Sprintf("unsupported '%s' patch value", userNameAttr), "value", op.Value)
+				return scim.Resource{}, errors.ScimErrorBadParams([]string{fmt.Sprintf("%v", op)})
 			}
-			user.Active = &active
+			user.UserName = userName
 		case op.Path.String() == activeAttr:
+			if op.Value == nil {
+				user.Active = nil
+				continue
+			}
 			active, ok := op.Value.(bool)
 			if !ok {
-				level.Error(u.logger).Log("msg", "unsupported 'active' patch value", "value", op.Value)
+				level.Error(u.logger).Log("msg", fmt.Sprintf("unsupported '%s' patch value", activeAttr), "value", op.Value)
 				return scim.Resource{}, errors.ScimErrorBadParams([]string{fmt.Sprintf("%v", op)})
 			}
 			user.Active = &active
+		case op.Path.String() == nameAttr+"."+givenNameAttr:
+			givenName, ok := op.Value.(string)
+			if !ok {
+				level.Info(u.logger).Log("msg", fmt.Sprintf("unsupported '%s' patch value", nameAttr+"."+givenNameAttr), "value", op.Value)
+				return scim.Resource{}, errors.ScimErrorBadParams([]string{fmt.Sprintf("%v", op)})
+			}
+			user.GivenName = &givenName
+		case op.Path.String() == nameAttr+"."+familyNameAttr:
+			familyName, ok := op.Value.(string)
+			if !ok {
+				level.Info(u.logger).Log("msg", fmt.Sprintf("unsupported '%s' patch value", nameAttr+"."+familyNameAttr), "value", op.Value)
+				return scim.Resource{}, errors.ScimErrorBadParams([]string{fmt.Sprintf("%v", op)})
+			}
+			user.FamilyName = &familyName
+		case op.Path.String() == nameAttr:
+			err = u.patchName(op.Value, op, user)
+			if err != nil {
+				return scim.Resource{}, err
+			}
+		case op.Path.String() == emailsAttr:
+			err = u.patchEmails(op.Value, op, user)
+			if err != nil {
+				return scim.Resource{}, err
+			}
+		case op.Path.AttributePath.String() == emailsAttr:
+			attrExpression, ok := op.Path.ValueExpression.(*filter.AttributeExpression)
+			if !ok {
+				level.Info(u.logger).Log("msg", "unsupported patch path", "path", op.Path)
+				return scim.Resource{}, errors.ScimErrorBadParams([]string{fmt.Sprintf("%v", op)})
+			}
+			// Only matching by email type (work, etc.) is supported.
+			if attrExpression.AttributePath.String() != typeAttr || attrExpression.Operator != filter.EQ {
+				level.Info(u.logger).Log("msg", "unsupported patch path", "path", op.Path, "expression", attrExpression.AttributePath.String())
+				return scim.Resource{}, errors.ScimErrorBadParams([]string{fmt.Sprintf("%v", op)})
+			}
+			emailType, ok := attrExpression.CompareValue.(string)
```

Comment:
Those type assertions checks are making it hard to follow the logic, I think maybe something like a single call once you know the attribute's type (that is, once `op.Path.AttributePath.String() == emailsAttr` is determined), calling a helper like `attr, err := toEmailsAttr(op.Path)` and then having a strongly-typed struct for the rest of the logic would improve readability, with a single error-returning path if any of the type-checks don't work.